### PR TITLE
`Agent()` tool calling loop + `Network()` ReAct impl.

### PR DESCRIPTION
**Pull request by charlypoly**

### CODE REVIEW

File: demo/inngest.ts [hunk 1]
```diff
@@ -24,7 +25,7 @@ export const fn = inngest.createFunction(
   { id: "agent" },
   { event: "agent/run" },
   async ({ event, step }) => {
-    const model = anthropic({ model: "claude-3-5-haiku-latest", max_tokens: 1024, step });
+    const model = openai({ model: "gpt-4", step });
```
Comment (charlypoly) [inline demo/inngest.ts:28]: temporary, for the sake of local testing and iterations

File: demo/inngest.ts [hunk 2]
```diff
@@ -33,36 +34,35 @@ export const fn = inngest.createFunction(
...
-      if (network.state.kv.has("files")) {
-        // Okay, we have some files.  Did an agent run tests?
-        return executingAgent;
-      }
-
-      return defaultRoutingAgent.withModel(model);
-    });
-
-    return result;
+    // // 2. A network of agents that works together
```
Comment (charlypoly) [inline demo/inngest.ts:37]: same, will work on this part tomorrow

File: package.json [hunk 3]
```diff
@@ -7,7 +7,8 @@
     "changeset": "changeset",
     "build": "rm -rf dist && tsc",
     "lint": "eslint .",
-    "release": "pnpm run build && changeset publish"
+    "release": "pnpm run build && changeset publish",
+    "dev": "nodemon --watch demo -e ts --exec \"ts-node -P tsconfig-demo.json demo/index.ts\""
```
Comment (charlypoly) [inline package.json:11]: temporary crap to make the demo work with the local AgentKit library

File: src/adapters/openai.ts [hunk 4]
```diff
@@ -19,14 +66,31 @@ export const requestParser: AgenticModel.RequestParser<OpenAi.AiModel> = (
 ) => {
   const request: AiAdapter.Input<OpenAi.AiModel> = {
     messages: messages.map((m) => {
+      const role = StateRoleToOpenAiRole[m.role];
       return {
-        role: m.role,
+        ...(m.stop_reason ? { finish_reason: StateStopReasonToOpenAiStopReason[m.stop_reason] } : {}),
+        role,
         content: m.content,
+        // NOTE: this is very ugly, we need to better handle different shape of messages
```
Comment (charlypoly) [inline src/adapters/openai.ts:74]: @jpwilliams I wouldn't say I like this current state; too much error-prone IMO

We could extract this in a helper that transforms based on the `role`

File: src/agent.ts [hunk 5]
```diff
@@ -108,6 +109,33 @@ export class Agent {
       history = modified.history;
     }
 
+    let hasMoreActions = true
+
+    do {
+      const inference = await this.performInference(input, p, history, system, network);
```
Comment (charlypoly) [inline src/agent.ts:115]: I guess we can add additional lifecycle callback here

File: src/agent.ts [hunk 6]
```diff
@@ -174,19 +203,17 @@ export class Agent {
...
-          continue;
-        }
+        // TODO: handle error and send them back to the LLM
 
         output.push({
           role: "tool_result",
-          content: {
-            type: "tool_result",
+          tools: [{
+            type: "tool",
```
Comment (charlypoly) [inline src/agent.ts:211]: Note: this `type` is likely to be an issue with Anthropic

File: src/network.ts [hunk 7]
```diff
@@ -233,23 +235,23 @@ export const defaultRoutingAgent = new Agent({
             .describe("The name of the agent that should handle the request"),
         })
         .strict(),
-      handler: (input, { agent, network }) => {
```
Comment (charlypoly) [inline src/network.ts:236]: `agent` was always equal to the `defaultAgenticRouter`

File: src/agent.ts [hunk 8]
```diff
@@ -86,15 +86,20 @@ export class Agent {
    */
   async run(
     input: string,
-    { model, network }: Agent.RunOptions | undefined = {},
+    {
+      model,
+      network,
+      maxIter = this.tools.size + 1,
```
Comment (tonyhb) [inline src/agent.ts:92]: Are we expecting _every_ tool to be called _exactly_ once here?  This seems really hard to tell up front 🤔

File: src/agent.ts [hunk 9]
```diff
@@ -108,6 +113,39 @@ export class Agent {
       history = modified.history;
     }
 
+    let hasMoreActions = true;
+    let iter = 0;
+    do {
+      const inference = await this.performInference(
```
Comment (tonyhb) [inline src/agent.ts:119]: How do lifecycles work here?  With looping, this breaks lifecycles.  The "before call" and "before inference" calls are no longer doing what we expect.

**Pull request merged**